pkg/sql/catalog/descriptor.go

```diff
@@ -632,63 +632,75 @@ type TableDescriptor interface {
-	GetUniqueWithoutIndexConstraints() []descpb.UniqueWithoutIndexConstraint
-	// AllActiveAndInactiveUniqueWithoutIndexConstraints returns all unique
-	// constraints that are not enforced by an index, including both "active"
-	// ones on the table descriptor which are being enforced for all writes, and
-	// "inactive" ones queued in the mutations list.
-	AllActiveAndInactiveUniqueWithoutIndexConstraints() []*descpb.UniqueWithoutIndexConstraint
-
-	// ForeachOutboundFK calls f for every outbound foreign key in desc until an
-	// error is returned.
-	ForeachOutboundFK(f func(fk *descpb.ForeignKeyConstraint) error) error
-	// ForeachInboundFK calls f for every inbound foreign key in desc until an
-	// error is returned.
-	ForeachInboundFK(f func(fk *descpb.ForeignKeyConstraint) error) error
-	// AllActiveAndInactiveForeignKeys returns all foreign keys, including both
-	// "active" ones on the index descriptor which are being enforced for all
-	// writes, and "inactive" ones queued in the mutations list. An error is
-	// returned if multiple foreign keys (including mutations) are found for the
-	// same index.
-	AllActiveAndInactiveForeignKeys() []*descpb.ForeignKeyConstraint
-
-	// GetChecks returns information about this table's check constraints, if
-	// there are any. Only valid if IsTable returns true.
-	GetChecks() []*descpb.TableDescriptor_CheckConstraint
-	// AllActiveAndInactiveChecks returns all check constraints, including both
-	// "active" ones on the table descriptor which are being enforced for all
-	// writes, and "inactive" new checks constraints queued in the mutations list.
-	// Additionally,  if there are any dropped mutations queued inside the mutation
-	// list, those will not cancel any "active" or "inactive" mutations.
-	AllActiveAndInactiveChecks() []*descpb.TableDescriptor_CheckConstraint
-	// ActiveChecks returns a list of all check constraints that should be enforced
-	// on writes (including constraints being added/validated). The columns
-	// referenced by the returned checks are writable, but not necessarily public.
-	ActiveChecks() []descpb.TableDescriptor_CheckConstraint
+	// FindConstraintWithName is like FindConstraintWithID but for names.
+	FindConstraintWithName(name string) (Constraint, error)
+
+	// CheckConstraintColumns returns the slice of columns referenced by a check
+	// constraint.
+	CheckConstraintColumns(ck CheckConstraint) []Column
+	// ForeignKeyReferencedColumns returns the slice of columns referenced by an
+	// inbound foreign key.
+	ForeignKeyReferencedColumns(fk ForeignKeyConstraint) []Column
+	// ForeignKeyOriginColumns returns the slice of columns originating in this
+	// table for an outbound foreign key.
+	ForeignKeyOriginColumns(fk ForeignKeyConstraint) []Column
+	// UniqueWithoutIndexColumns returns the slice of columns which are
+	// defined as unique by a non-index-backed constraint.
+	UniqueWithoutIndexColumns(uwoi UniqueWithoutIndexConstraint) []Column
```

Comment:
Why do you think those methods belong here? It seems to me more appropriately to define them just as function that takes in a (ck|fk|uwoi) and return some property of that input constraint. Do we need anything from the table to serve those questions?

Reply:
Good question. The table is required to map the column ID to a `catalog.Column`, and I really prefer using the latter instead of IDs whenever possible. 

See the similar `[]Column`-returning methods taking indexes. These can be quite powerful and removing this layer of indirection of the column ID makes for some more expressive code at the call sites. See informationSchemaConstraintColumnUsageTable` for an example. Arguably some of them are never called, but if they're there it's because they used to, and perhaps will be again.